File: day02.py
"""Solution to day 2 of Advent of Code"""

from get_input import get_input, line_parser


def part1(lines):
    position = 0+0j
    diffs = {
        'up': 0-1j,
        'down': 0+1j,
        'forward': 1+0j,
    }
    for (command, size) in lines:
        position += diffs[command] * size
    return int(position.real) * int(position.imag)


def part2(lines):
    position = 0+0j
    aim = 0
    for (command, size) in lines:
        if command == "up":
            aim -= size
        elif command == 'down':
            aim += size
        elif command == 'forward':
            position += complex(size, aim * size)
        else:
            raise NotImplementedError
    return int(position.real) * int(position.imag)


def parse(line):
    command, depth = line.split(' ')
    return (command, int(depth))


if __name__ == "__main__":
    LINES = line_parser(get_input(day=2, year=2021), parse=parse)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day02.rs
use common::get_input;
use common::point::Point2D;

type Input = Vec<Command>;

#[derive(Debug, Clone)]
enum Command {
    Up(isize),
    Down(isize),
    Forward(isize),
}

fn part1(input: &Input) -> usize {
    let mut position: Point2D<isize> = Point2D::new(0, 0);
    for command in input {
        position += match command {
            Command::Up(v) => Point2D::new(0, -v),
            Command::Down(v) => Point2D::new(0, *v),
            Command::Forward(v) => Point2D::new(*v, 0),
        }
    }
    return (position.x * position.y) as usize
}

fn part2(input: &Input) -> usize {
    let mut position: Point2D<isize> = Point2D::new(0, 0);
    let mut aim: isize = 0;
    for command in input {
        match command {
            Command::Up(v) => aim -= v,
            Command::Down(v) => aim += v,
            Command::Forward(v) => position += Point2D::new(*v, aim * v),
        }
    }
    return (position.x * position.y) as usize
}

fn parse(text: String) -> Input {
    text.trim()
        .split('\n')
        .map(|t| {
            let values: Vec<&str> = t.split(' ').collect();
            let value = values[1].parse::<isize>().unwrap();
            match values[0] {
                "up" => Command::Up(value),
                "down" => Command::Down(value),
                "forward" => Command::Forward(value),
                _ => unimplemented!()
            }
        })
        .collect()
}

fn main() {
    let input = parse(get_input(02, 2021));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
